pkg/storage/cacher.go

```diff
@@ -307,21 +381,53 @@ func (c *Cacher) Codec() runtime.Codec {
 	return c.storage.Codec()
 }
 
+func (c *Cacher) indexValues(event *watchCacheEvent) ([]string, bool) {
+	if c.indexFunc == nil {
+		return nil, false
+	}
+	// TODO: Currently we assume that only one index exists, thus
+	// exactly 1 value will be returned.
+	matchValues := c.indexFunc(event.Object)
+	if event.PrevObject == nil {
+		return []string{matchValues[0].Value}, true
+	}
+	prevMatchValues := c.indexFunc(event.Object)
+	if matchValues[0].Value == prevMatchValues[0].Value {
+		return []string{matchValues[0].Value}, true
+	}
+	return []string{matchValues[0].Value, prevMatchValues[0].Value}, true
+}
+
 func (c *Cacher) processEvent(event watchCacheEvent) {
+	indexValues, supported := c.indexValues(&event)
+
 	c.Lock()
 	defer c.Unlock()
-	for _, watcher := range c.watchers {
+	// Iterate over "allWatchers" no matter what the indexing function is.
+	for _, watcher := range c.watchers.allWatchers {
 		watcher.add(event)
 	}
+	if supported {
+		// Iterate over watchers interested in the given values of the index.
+		for _, indexValue := range indexValues {
+			for _, watcher := range c.watchers.valueWatchers[indexValue] {
+				watcher.add(event)
+			}
+		}
+	} else {
+		// Iterate over watchers interested in exact values for all values.
```

Comment:
Yes - it is executed. Basically if the index doesn't return a value, that means, that potentially objects with any value of that field can be filtered out.
As an example, if your filter accepts everything, then "supported" would be false (because basically no values will be returned from the FilterIndex function).